apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -148,27 +155,61 @@ impl ApplyToError {
     }
 }
 
+// Rust doesn't allow implementing methods directly on tuples like
+// (Option<JSON>, Vec<ApplyToError>), so we define a trait to provide the
+// methods we need, and implement the trait for the tuple in question.
+pub(super) trait ApplyToResultMethods {
+    fn prepend_errors(self, errors: Vec<ApplyToError>) -> Self;
+
+    fn and_then_collecting_errors(
+        self,
+        f: impl FnOnce(&JSON) -> (Option<JSON>, Vec<ApplyToError>),
+    ) -> (Option<JSON>, Vec<ApplyToError>);
+}
+
+impl ApplyToResultMethods for (Option<JSON>, Vec<ApplyToError>) {
+    // Intentionally taking ownership of self to avoid cloning, since we pretty
+    // much always use this method to replace the previous (value, errors) tuple
+    // before returning.
+    fn prepend_errors(self, mut errors: Vec<ApplyToError>) -> Self {
```

Comment:
It's interesting that this implementation of `ApplyToResultMethods::prepend_errors` can take the owned `errors` vector as `mut`, even though the trait declaration doesn't say anything about that.

Reply:
This is because the method takes ownership of the vector, so it's only mutable in the context of the function!